### PR TITLE
Generate API documentation for service resource functions

### PR DESCRIPTION
**Pull request by aneeshafedo**
## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: openapi-cli/src/test/resources/expected/petstore.bal [hunk 1]
```diff
@@ -3,10 +3,27 @@ import ballerina/http;
 listener http:Listener ep0 = new (80, config = {host: "petstore.openapi.io"});
 
 service /v1 on ep0 {
+    # List all pets
+    #
+    # + 'limit - How many items to return at one time (max 100)
+    # + return - returns can be any of following types
+    # Pets (An paged array of pets)
```
Comment (aneeshafedo) [inline openapi-cli/src/test/resources/expected/petstore.bal:10]: @NipunaRanasinghe @lnash94 Due to the issue https://github.com/ballerina-platform/ballerina-lang/issues/29607, this multi line return description is not formatted correctly.

Comment (NipunaRanasinghe) [reply openapi-cli/src/test/resources/expected/petstore.bal:10]: Ack

File: openapi-core/src/main/java/io/ballerina/openapi/core/GeneratorUtils.java [hunk 2]
```diff
@@ -272,6 +274,17 @@ private static void extractPathParameterDetails(Operation operation, List<Node>
...
                 functionRelativeResourcePath.add(createToken(SLASH_TOKEN));
+
+                // Add documentation
+                if (resourceFunctionDocs != null) {
+                    String parameterName = paramName.text();
+                    String paramComment = parameter.getDescription() != null && !parameter.getDescription().isBlank() ?
+                            parameter.getDescription() : DEFAULT_PARAM_COMMENT;
+                    MarkdownParameterDocumentationLineNode paramAPIDoc =
+                            DocCommentsGenerator.createAPIParamDoc(parameterName
+                                    , paramComment);
```
Comment (NipunaRanasinghe) [inline openapi-core/src/main/java/io/ballerina/openapi/core/GeneratorUtils.java:285]:
```suggestion
                            DocCommentsGenerator.createAPIParamDoc(parameterName , paramComment);
```

File: openapi-core/src/main/java/io/ballerina/openapi/core/generators/service/BallerinaServiceGenerator.java [hunk 3]
```diff
@@ -293,11 +320,31 @@ private FunctionDefinitionNode getResourceFunction(Map.Entry<PathItem.HttpMethod
...
+        } else if (operation.getValue().getDescription() != null && !operation.getValue().getDescription().isBlank()) {
+            resourceFunctionDocs.addAll(DocCommentsGenerator.createAPIDescriptionDoc(
+                    operation.getValue().getDescription(), true));
+        } else {
+            resourceFunctionDocs.addAll(DocCommentsGenerator.createAPIDescriptionDoc(
+                    DEFAULT_FUNC_COMMENT, true));
+        }
+    }
+
+
```
Comment (NipunaRanasinghe) [inline openapi-core/src/main/java/io/ballerina/openapi/core/generators/service/BallerinaServiceGenerator.java:347]: Additional blank line

File: openapi-core/src/main/java/io/ballerina/openapi/core/generators/service/ParametersGenerator.java [hunk 4]
```diff
@@ -137,6 +142,19 @@ public void generateResourcesInputs(Map.Entry<PathItem.HttpMethod, Operation> op
                         }
                     }
                 }
+
+                if (param != null) {
+                    String parameterName = param instanceof RequiredParameterNode ?
+                            ((RequiredParameterNode) param).paramName().get().text() :
+                            ((DefaultableParameterNode) param).paramName().get().text();
```
Comment (NipunaRanasinghe) [inline openapi-core/src/main/java/io/ballerina/openapi/core/generators/service/ParametersGenerator.java:149]: Casting to `DefaultableParameterNode` in else case might not be safe in here if there will be be rest parameters. Shall we make sure that rest parameters are not allowed for resource functions?

File: openapi-core/src/main/java/io/ballerina/openapi/core/generators/service/ReturnTypeGenerator.java [hunk 5]
```diff
@@ -122,31 +130,67 @@ public ReturnTypeGenerator(BallerinaTypesGenerator ballerinaSchemaGenerator, Str
...
+                        values[0], values[1]));
+            }
+            String dummyTypeWithDescription = returnDescriptionForUnions.toString() + "type a A;";
+            ModuleMemberDeclarationNode moduleMemberDeclarationNode = NodeParser.parseModuleMemberDeclaration(
+                    dummyTypeWithDescription);
+            TypeDefinitionNode typeDefinitionNode = (TypeDefinitionNode) moduleMemberDeclarationNode;
+            MetadataNode metadataNode = typeDefinitionNode.metadata().get();
+            MarkdownDocumentationNode returnDoc = (MarkdownDocumentationNode) metadataNode.children().get(0);
+            resourceFunctionDocs.add(returnDoc);
+
```
Comment (NipunaRanasinghe) [inline openapi-core/src/main/java/io/ballerina/openapi/core/generators/service/ReturnTypeGenerator.java:187]: additional blank line

File: openapi-core/src/main/java/io/ballerina/openapi/core/generators/service/ParametersGenerator.java (hunk 4 above)
Comment (aneeshafedo) [reply openapi-core/src/main/java/io/ballerina/openapi/core/generators/service/ParametersGenerator.java:149]: According to the implementation here in openapi-to-ballerina-service generation, any parameter given in the openapi has mapped to either required parameter of defaultable  parameter. AFAIK in the current implementation there is no openapi parameters schema represented as rest type parameter in Ballerina. 

@lnash94 appreciate you thoughts too.

File: openapi-core/src/main/java/io/ballerina/openapi/core/generators/service/BallerinaServiceGenerator.java (hunk 3 above)
Comment (aneeshafedo) [reply openapi-core/src/main/java/io/ballerina/openapi/core/generators/service/BallerinaServiceGenerator.java]: Updated

File: openapi-core/src/main/java/io/ballerina/openapi/core/generators/service/ReturnTypeGenerator.java (hunk 5 above)
Comment (aneeshafedo) [reply openapi-core/src/main/java/io/ballerina/openapi/core/generators/service/ReturnTypeGenerator.java]: Updated

File: openapi-core/src/main/java/io/ballerina/openapi/core/generators/service/ParametersGenerator.java (hunk 4 above)
Comment (lnash94) [reply openapi-core/src/main/java/io/ballerina/openapi/core/generators/service/ParametersGenerator.java:149]: Only required and default parameters are mapped in openapi-to-ballerina service/client generations. Although Resource functions can have rest parameters as the path parameters in ballerina but still need to investigate a representation for the rest parameters in OpenAPI specification under the parameter schemas.

File: openapi-core/src/main/java/io/ballerina/openapi/core/generators/service/ReturnTypeGenerator.java [hunk 6]
```diff
@@ -272,39 +315,47 @@ private ReturnTypeDescriptorNode getReturnNodeForSchemaType(Set<Map.Entry<String
...
                 } else {
                     SimpleNameReferenceNode node = createReturnTypeInclusionRecord(code, bodyType);
-                    qualifiedNodes.add(node.name().text());
+                    typeName = node.name().text();
+                }
+            }
+            if (typeName != null) {
+                qualifiedNodes.add(typeName);
+                if (response.getValue().getDescription() != null && !response.getValue().getDescription().isEmpty()) {
+                    returnDescription.add(typeName.trim() + "|" + response.getValue().getDescription().trim());
```
Comment (lnash94) [inline openapi-core/src/main/java/io/ballerina/openapi/core/generators/service/ReturnTypeGenerator.java:358]: Shall we reuse the constant value for the pipeline "|" if we defined it already?

Comment (aneeshafedo) [reply openapi-core/src/main/java/io/ballerina/openapi/core/generators/service/ReturnTypeGenerator.java]: Updated

**Review by NipunaRanasinghe — APPROVED** [review]
LGTM

**Pull request merged**